README.md

```diff
@@ -89,58 +92,275 @@ Now you can import like this:
+
+    var bench = try zbench.Benchmark.init(second, bench_iterations, gpa.allocator());
+    defer bench.deinit();
 }
 ```
+The first argument to `init` is an estimate in nanoseconds for the maximum amount of time we are willing to wait for any given benchmark to finish,
+the second argument is the maximum number of repetitions or runs for each benchmark. Now to perform the benchmark we pass `myBenchRunner` to `bench.run`,
+and the name of our benchmark. The complete example looks like this:
+```zig
+const std = @import("std");
+const zbench = @import("zbench");
 
-### Compatibility Notes
-Zig is in active development and the APIs can change frequently, making it challenging to support every dev build. This project currently aims to be compatible with stable, non-development builds to provide a consistent experience for the users.
+fn myBenchRunner() void {
+    // Code to benchmark here
+}
 
-***Supported Version***: As of now, zBench is tested and supported on Zig version ***0.11.0***.
+pub fn main() !void {
+    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
 
-### Benchmark Functions
-Benchmark functions have the following signature:
+    const second: u64 = 1_000_000_000;
+    const bench_iterations: u64 = 128;
 
+    var bench = try zbench.Benchmark.init(second, bench_iterations, gpa.allocator());
+    defer bench.deinit();
+
+    const bench_result = try bench.run(myBenchRunner, "Hello bench");
+    try bench_result.prettyPrint(true);
+}
+```
+`bench.run` returns a `BenchmarkResult` which can be pretty-printed (more parsing functionality may be added in the future):
+```yaml
+benchmark                 runs     time (avg ± σ)         (min ............. max)      p75        p99        p995
+---------------------------------------------------------------------------------------------------------------------
+Hello bench               128      8.806µs ± 67.0ns       (8.351µs ... 9.185µs)        8.813µs    9.31µs     9.185µs
+```
+If `myBenchRunner` needs to allocate we could have declared `std.mem.Allocator` as a parameter:
 ```zig
-fn(b: *zbench.Benchmark) void
+fn myBenchRunner(alloc: std.mem.Allocator) void {
+    // Code to benchmark here
+}
 ```
-The function body contains the code you wish to benchmark.
+### Aggregate runner
+In most cases we want to first set up some state relevant to the benchmark, but we aren't interested in benchmarking the state setup code.
+You can use an aggregate runner (ie. struct) to split such initialisation from run code:
+```zig
+const std = @import("std");
+const zbench = @import("zbench");
+
+const StructRunner = struct {
+    const Self = @This();
+
+    nums: [10]u64,
+
+    pub fn init(_: std.mem.Allocator) !Self {
+        return Self { .nums = .{1, 2, 3, 4, 5, 6, 7, 8, 9, 10} };
+    }
+
+    pub fn run(self: *Self) void {
+        var i: usize = 1;
+        while (i < self.nums.len) : (i += 1) {
+            self.nums[i] += self.nums[i-1];
+        }
+    }
```

Comment:
We might want to mention that if you benchmark code that has basically a compile-time-known result, you have to be careful with optimizations (-Doptimize=...). An option for such cases would be to use `std.mem.doNotOptimizeAway([result])`

Reply:
Yeah that's true. Ooh I didn't know about that function, that is useful! I was doing a lot of acrobatics in my own benchmarks to prevent them from being optimized away :disappointed: